aleph-client/src/contract/mod.rs

```diff
@@ -0,0 +1,144 @@
+/// Utilities for writing contract wrappers.
+///
+/// The functions in this module simplify parsing the data types returned by contract calls.
+pub mod util;
+
+use std::{
+    fmt::{Debug, Formatter},
+    fs::File,
+};
+
+use ac_primitives::ExtrinsicParams;
+use anyhow::{anyhow, Context, Result};
+use contract_metadata::ContractMetadata;
+use contract_transcode::{ContractMessageTranscoder, Value};
+use ink_metadata::{InkProject, MetadataVersioned};
+use serde_json::{from_reader, from_str, from_value, json};
+use sp_core::{crypto::AccountId32, Pair};
+use substrate_api_client::{compose_extrinsic, GenericAddress, XtStatus};
+
+use crate::{try_send_xt, AnyConnection, Connection, SignedConnection};
+
+/// Represents a contract instantiated on the chain.
+pub struct ContractInstance {
+    address: AccountId32,
+    ink_project: InkProject,
+}
+
+impl ContractInstance {
+    const MAX_READ_GAS: u64 = 500000000000u64;
+    const MAX_GAS: u64 = 10000000000u64;
+
+    /// Creates a new contract instance under `address` with metadata read from `metadata_path`.
+    pub fn new(address: AccountId32, metadata_path: &str) -> Result<Self> {
+        Ok(Self {
+            address,
+            ink_project: load_metadata(metadata_path)?,
+        })
+    }
+
+    /// The address of this contract instance.
+    pub fn address(&self) -> &AccountId32 {
+        &self.address
+    }
+
+    /// The metadata of this contract instance.
+    pub fn ink_project(&self) -> &InkProject {
+        &self.ink_project
+    }
+
+    /// Reads the value of a read-only, 0-argument call via RPC.
+    pub fn contract_read0(&self, conn: &Connection, message: &str) -> Result<Value> {
+        self.contract_read(conn, message, &[])
+    }
+
+    /// Reads the value of a read-only call via RPC.
+    pub fn contract_read(&self, conn: &Connection, message: &str, args: &[&str]) -> Result<Value> {
+        let payload = self.encode(message, args)?;
+        let request = self.contract_read_request(&payload);
+        let response = conn
+            .get_request(request)?
+            .context("RPC request error - there may be more info in node logs.")?;
+        let response_data = from_str::<serde_json::Value>(&response)?;
+        let hex_data = response_data["result"]["Ok"]["data"]
+            .as_str()
+            .context("Contract response data not found - the contract address might be invalid.")?;
+        self.decode_response(message, hex_data)
+    }
+
+    /// Executes a contract call.
+    pub fn contract_exec(
```

Comment:
very weak nit: we have only `contract_exec`, but regarding reading, we have `contract_read` and `contract_read0`; I would encourage you to add `contract_exec0`, since in tests we are very likely to have simple non-arg messages to trigger (like `flipper`)